implement renewal of pki certs with min_seconds_remaining argument

joemiller opened the pull request:
This implements renewal for certificates in the `vault_pki_secret_backend_cert` and `vault_pki_secret_backend_sign` resources.  Related issue: https://github.com/terraform-providers/terraform-provider-vault/issues/353

Some notes on the implementation:

* Adds a new optional argument `min_seconds_remaining`. The  default is 7 days (604800s). If the certificate's `expiration` is within this many seconds a new certificate will be fetched from Vault. The reason I chose to go with seconds instead of days was to cover cases where short-lived (perhaps less than a day) certificates are desired. I suppose "hours" might be a reasonable choice here as well. Days would be too long.

* Adds a new computed attribute `expiration` to these resources. This value was already being returned by the Vault API and now it will be stored in TF state.

* Based loosely off of the way in which the acme TF provider handles cert refresh, except they use `min_days_remaining`. This seems fine for certs issued by public CAs but some internal CAs may desire shorter expiration - https://github.com/terraform-providers/terraform-provider-acme/blob/0077b51dadc8b9fc4485e5b2a2844dedb063d9ae/acme/resource_acme_certificate.go#L83-L109

* I bumped the `mitchellh/copystructure` and `mitchellh/reflectwalk` vendor'd deps to `v1.0.0`. I ran into a bug `panic: interface conversion: interface {} is schema.schemaMap, not *schema.schemaMap` when implementing the custom Diff funcs and this was the resolution.

joemiller commented:
tests are passing now. The issue was that the test case's pki role has a 1 hour max ttl, but min_seconds_remaining defaults to 7days. I added `min_seconds_remaining = 60` to the test so that it would not immediately be marked for renewal

joemiller commented:
Any feedback on this PR? thanks

tyrannosaurus-becks commented on vault/resource_pki_secret_backend_cert.go:
```diff
@@ -192,16 +205,47 @@ func pkiSecretBackendCertCreate(d *schema.ResourceData, meta interface{}) error
+
+	renewTime := expireTime.Add(-time.Duration(minSeconds) * time.Second)
+	if time.Now().After(renewTime) {
+		log.Printf("[DEBUG] certificate %q is due for renewal, expires %s, renewal time %s", d.Id(), expireTime, renewTime)
```
It would be great to get some test coverage of this code path. I see that the rest of the code in this method is currently hit, but not in here.

joemiller replied:
ok, i'll work on that

tyrannosaurus-becks commented on vault/resource_pki_secret_backend_cert.go:
```diff
@@ -192,16 +205,47 @@ func pkiSecretBackendCertCreate(d *schema.ResourceData, meta interface{}) error
 func pkiSecretBackendCertUpdate(d *schema.ResourceData, m interface{}) error {
+	// If the certificate or private_key have been marked as changed by our custom diff function we need
+	// to create a new certificate and key
+	if d.HasChange("certificate") || d.HasChange("private_key") {
```
This code path doesn't appear to have test coverage yet. Would be great to get some here too.

tyrannosaurus-becks commented on vault/resource_pki_secret_backend_sign.go:
```diff
@@ -194,16 +206,46 @@ func pkiSecretBackendSignCreate(d *schema.ResourceData, meta interface{}) error
+	}
+
+	log.Printf("[DEBUG] certificate %q is not due for renewal, expires %s, renewal time %s", d.Id(), expireTime, renewTime)
+	return nil
```
It would be great to get test coverage on all the code changes in this file too - the new fields that have been added, and the certificate that's generated.

joemiller commented:
@tyrannosaurus-becks I am happy to make this opt-in for backwards compatibility. i'll work on the requested changes

joemiller commented:
@tyrannosaurus-becks I've added an `auto_renew = [true|false] (default: false)` attribute to both the `vault_pki_secret_backend_sign` and `vault_pki_secret_backend_cert` resources  to aid with backward compat, as requested. Also added acceptance tests covering renewal. Though I notice a switch to circleci is in progress and unlike the previous travis tests it's not running the acceptance tests currently.

I also include a commit with some doc fixes. 1) cleaned up some trailing whitespace and 2) "ttl - time to leave" I am assuming should be "time to live" per the related vault docs

joemiller commented:
updated to remove vendor/vendor.json which i noticed i had accidentally re-added in my most recent rebase from master
